### DISCUSSION TITLE
Bug: #15961 was incorrectly closed. There is an error in eslint for node.js public class variables

### DISCUSSION THREAD

**Discussion by TotallyInformation** [eslint/eslint#15965] [General] [answered]
### Environment

Node version: v12.13.1
npm version: N/A
Local ESLint version: 8.16.0
Global ESLint version: 8.16.0
Operating System: Windows 10

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

See https://github.com/eslint/eslint/issues/15961
</details>

```
class UibPackages {
    // TODO: Replace _XXX with #XXX once node.js v14 is the minimum supported version
    /** Flag to indicate whether setup() has been run (ignore the false eslint error)
     * @type {boolean}
     * @protected 
     */
    _isConfigured = false

    /** @type {Array<string>} Updated by updateMergedPackageList which is called first in setup and then in various updates */
    mergedPkgMasterList = []
    /** @type {string} The name of the package.json file 'package.json' */
    packageJson = 'package.json'
    /** @type {*} The uibRoot package.json contents */
    uibPackageJson
    /** @type {string} Get npm's global install location */
    globalPrefix // set in constructor

    constructor() {
      this.fred = 'hi there'
      console.log(this.packageJson)
    }
    
    showMe() {
      console.log(this.packageJson, this.fred)
    }
}

const x = new UibPackages()

console.log("Hello, World!", x.showMe(), x.packageJson);
```

### What did you expect to happen?

In node.js, public class variables were supported in v12+, eslint does not allow them until ECMA 13 which is incorrect. They should be supported from at least ECMA 10 if not 9 for node.js.

Please don't blindly close this as with #15961, I have tested this and it works fine in node.js v12.13.1 as shown by https://onecompiler.com/nodejs/3y5zhpz3d

Public class variables were, as I said, introduced in node.js v12. ESLINT should not give up with an undismissable error for something that has been supported for so long. How can I get rid of the error - how can I turn off this unnecessary error please?

### What actually happened?

Please see the incorrectly closed #15961 which contains the details of the error that is shown.

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

https://github.com/eslint/eslint/issues/15961

**Comment by mdjermanovic** [comment]
`parserOptions.ecmaVersion` is a version of the ECMAScript Language Specification.

Class fields are added in ECMAScript 2022 Language Specification: https://tc39.es/ecma262/2022/

They didn't exist in ECMAScript 2021 Language Specification: https://tc39.es/ecma262/2021/

You can check which feature was added in which version (the "Expected Publication Year" column) here:

https://github.com/tc39/proposals/blob/main/finished-proposals.md

Some environments may support features that have not yet been added to the language specification. If your target environments support ECMAScript 2022, you can set `parserOptions: { ecmaVersion: 2022 }`.

**Reply by mdjermanovic** [reply]
Also, note that 6, 7, 8, 9, 10, 11, 12, 13 represent editions of the specification.

6 = 2015 (commonly referred to as ES6)
7 = 2016
...
12 = 2021
13 = 2022

`parserOptions: { ecmaVersion: 12 }` enables ECMAScript 2021 syntax. It isn't related to versions of Node.js.

**Comment by TotallyInformation** [comment]
Hi @mdjermanovic, many thanks for your reply.

While I agree that the spec for this feature wasn't ratified until the 2022 spec, it was in proposal stage from early 2019. I don't believe that Node.js (actually the Google V8 engine) supports ALL of the 2022 spec in v12.x so using that isn't an option I'm afraid.

The point I guess I should really have made is that this should be an OPTION like so many other features in eslint. Then it could be turned on when using a version of JavaScript such as V8 that supported it in Node.js and Chrome for so long.

As it stands, this does not appear to be possible and instead you get an error in the linting. Because the error is generic, you can't simply ignore it as it might be valid in other cases.

It means that ESLINT is effectively forcing some developers to not use a perfectly usable, sensible and useful feature just because ESLINT doesn't provide an option for this check to be turned off and it doesn't provide a specific error which would allow it at least to be trapped sensibly.

Would it not be possible to put in a specific check with a matching error? That way developers could choose for themselves whether to use it or not?

Alternatively, is there a setting that allows me to tell ESLINT to use a known good set of checks for Node.js v12? I already have multiple ESLINT configs because my package contains a mix of front-end and back-end code so this really wouldn't be a problem. It is only a problem right now because there is no workaround that lets me use ESLINT with Node.js v12 valid code without having an error.

Many thanks. J.

**Comment by aladdin-add** [comment, marked as answer]
you can check out the plugin [eslint-plugin-n](https://www.npmjs.com/package/eslint-plugin-n). it has a rule for such a purpose: https://github.com/weiran-zsd/eslint-plugin-node/blob/master/docs/rules/no-unsupported-features/es-syntax.md

also, [eslint-plugin-es](https://www.npmjs.com/package/eslint-plugin-es) can be helpful.

**Reply by TotallyInformation** [reply]
Ah, thank you so much for that. I was starting to think I'd have to go back to JSLint! A very helpful reply.
